### ISSUE TITLE
Feature request: Maintain indentation on line wrap

### ISSUE THREAD

**Issue by hudcap** [pyzo/pyzo#493]
When an indented line gets wrapped, the new lines should be indented accordingly.

I'd be happy to write this myself, but I don't fully understand how. If someone could point me in the right direction, I should be able to figure it out.
The code for line wrapping seems to be here: [https://github.com/pyzo/pyzo/blob/49d5c69caf9ab39d025617775d8f9438d17dccad/pyzo/codeeditor/extensions/appearance.py#L1243](url)
I also found this, which seems to be related, but the answer is too cryptic for me: [https://stackoverflow.com/questions/25185064/configure-wrapping-of-qtextedit-to-add-indent](url)

Please pardon my ignorance -- I've never dealt with PyQt before.

**Comment by almarklein** [comment]
I was not even aware that Qt could do this. Nice! The block styling can be applied (at least) via the [QTextCursos](http://doc.qt.io/qt-4.8/qtextcursor.html#setBlockFormat) and via the [QTextBlock](http://doc.qt.io/qt-4.8/qtextblock.html#blockFormat).

My best guess is to start somewhere around [here](https://github.com/pyzo/pyzo/blob/master/pyzo/codeeditor/highlighter.py#L106). `bd` is a `QTextBlock`, and the indentation is known at that point. Looking forward to a PR!

**Comment by hudcap** [comment]
I can't seem to get `setIndent()` to work, even with a constant.
I tried putting this at the end of `highlighter.py`:

```python
format = QtGui.QTextBlockFormat()
format.setIndent(2)

cb = self.currentBlock()
cursor = QtGui.QTextCursor(cb)
cursor.mergeBlockFormat(format)
```

This should have put in an indent of 80 (characters?) based on [this](https://doc.qt.io/qt-4.8/qtextdocument.html#indentWidth-prop), but I don't see any indent.
`setBlockFormat` had the same (lack of) results.

Could some other part of the code be blocking the indent display?

**Comment by almarklein** [comment]
I think that's in pixels. But its odd, since it does not change when zooming in/out, which means that even if this can be made to work, we'd have to also ensure a correct value of the document's indent width ...

I also had a quick go doing essentially `cb.blockFormat().setIndent(2)`, but it did not work either (and indeed cb.blockFormat().indent() != 2).

I don't think we use this feature elsewhere, but perhaps due to how Qt works, we implicitly override it? Hard to say for sure. This seems to be harder that it initially looked ;)

**Comment by hudcap** [comment]
That's odd -- I'm getting `cb.blockFormat().indent() == 2`, so it seems to be setting correctly for me.

**Comment by almarklein** [comment]
Oh, perhaps I did something else wrong then. Do you also *see* the indentation though?

**Comment by hudcap** [comment]
No, I didn't see any indentation.
I tried `setLeftMargin(80)`, and I don't see that either.
Could `blockFormat()` be "disabled?"

**Comment by almarklein** [comment]
We're using `setViewportMargins()` to set the margins: https://github.com/pyzo/pyzo/blob/master/pyzo/codeeditor/base.py#L870

Not sure if that somehow undoes other margins ...

**Comment by hudcap** [comment]
It seems that QPlainTextEdit is being used, and not QTextEdit.
Is this right?
Because according to [this](https://stackoverflow.com/a/2729177/8150768), it won't work with QPlainTextEdit.

I'm a little out of my depth here, since Qt is very new to me...

**Comment by almarklein** [comment]
Yes, QPlainTextEdit is used. I assumed that it would work, because it also works with TextBlocks. I guess not :/

**Comment by BrenBarn** [comment]
Is there a reason QPlainTextEdit is being used?  To my mind wrap on indent is so important a feature that it would be worth looking at other possibilities to get this functionality?

**Comment by almarklein** [comment]
QPlainTextEdit is by far the most obvious choice to base an editor component on. Except maybe QScintilla, which we used before but that was not pleasant. You could write something from scratch, but that's *a lot* of work.

**Comment by BrenBarn** [comment]
Hmmm, what was unpleasant about QScintilla?  I've been looking at using it for another project of my own so I'm curious what was problematic about it for PyZo.

**Comment by almarklein** [comment]
Its been a few years, but I remember horrible API, and poor Unicode support.
